[masonry] Add check if a style is a grid-lanes

stwrt opened the pull request:
#### 3a2b39baabee9acd924b9cc5c51db5a369d77202
<pre>
[masonry] Add check if a style is a grid-lanes
<a href="https://bugs.webkit.org/show_bug.cgi?id=302709">https://bugs.webkit.org/show_bug.cgi?id=302709</a>
<a href="https://rdar.apple.com/problem/164960515">rdar://problem/164960515</a>

Reviewed by Sammy Gill.

New method to check if an item is a grid-lanes display type without having to manually check if it is an inline or regular grid-lanes.

Using isDisplayFlexibleOrGridFormattingContextBox() in places, so we do not need to perform checks for each type.
These checks are getting more complex now that we have flex, grid, and grid-lands.

*  Source/WebCore/html/HTMLButtonElement.cpp
*  Source/WebCore/layout/layouttree/LayoutBox.cpp
*  Source/WebCore/layout/layouttree/LayoutBox.h
*  Source/WebCore/rendering/style/RenderStyle.h
*  Source/WebCore/rendering/style/RenderStyleInlines.h

Canonical link: <a href="https://commits.webkit.org/303225@main">https://commits.webkit.org/303225@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa18c7f6a48ffb0be8ecfa1153df96ae3541b38e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4160 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42636 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139178 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83530 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6667dbc1-b7de-40c7-8586-baeb9b7a6f59) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133539 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3923 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100663 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/31c68ffa-79fc-414a-b502-b42af00638c7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134615 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2967 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117947 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81428 "Failed to checkout and rebase branch from PR 54106") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2853 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/689 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82369 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111579 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36082 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141824 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3829 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36649 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109018 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3908 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3398 "Found 2 new test failures: imported/w3c/web-platform-tests/mediacapture-streams/overconstrained_error.https.html imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue/color_hex.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109153 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27660 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2929 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114224 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57040 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3883 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32649 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3715 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67217 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3975 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3844 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

fantasai reviewed (commented):
Ah, this would be a good idea. But I suggest maybe having separate functions for:
- is grid or grid-lanes (i.e. creates a type of grid formatting context)
- is a 2-axis grid (grid display type)
- is a 1-axis grid (grid-lanes display type)

sammygill commented on Source/WebCore/layout/layouttree/LayoutBox.cpp:
```diff
@@ -169,7 +169,7 @@ bool Box::establishesFlexFormattingContext() const
 bool Box::establishesGridFormattingContext() const
 {
-    return isGridBox();
+    return isGridFormattingContext();
```
IMO this function definition should just have the code from the body of `isGridFormattingContxt()`

stwrt replied:
I was trying to avoid duplicating code here, and was trying to match the behavior in the above functions.

sammygill reviewed (approved):
I'm assuming the places where we are using `isDisplayFlexibleOrGridFormattingContextBox()` is intentional since we're extending that behavior but might be worth calling that out in the commit message

stwrt committed:
[masonry] Add check if a style is a grid-lanes
https://bugs.webkit.org/show_bug.cgi?id=302709
rdar://problem/164960515

Reviewed by Sammy Gill.

New method to check if an item is a grid-lanes display type without having to manually check if it is an inline or regular grid-lanes.

Using isDisplayFlexibleOrGridFormattingContextBox() in places, so we do not need to perform checks for each type.
These checks are getting more complex now that we have flex, grid, and grid-lands.

*  Source/WebCore/html/HTMLButtonElement.cpp
*  Source/WebCore/layout/layouttree/LayoutBox.cpp
*  Source/WebCore/layout/layouttree/LayoutBox.h
*  Source/WebCore/rendering/style/RenderStyle.h
*  Source/WebCore/rendering/style/RenderStyleInlines.h

Canonical link: https://commits.webkit.org/303225@main
